include/deal.II/numerics/smoothness_estimator.templates.h

```diff
@@ -0,0 +1,434 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1998 - 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_smoothness_estimator_templates_h
+#define dealii_smoothness_estimator_templates_h
+
+#include <deal.II/base/quadrature.h>
+
+#include <deal.II/fe/fe_series.h>
+
+#include <deal.II/hp/q_collection.h>
+
+#include <deal.II/numerics/smoothness_estimator.h>
+
+#include <algorithm>
+#include <cmath>
+#include <functional>
+#include <utility>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+namespace SmoothnessEstimator
+{
+  namespace
+  {
+    /**
+     * Resizes @p coeff to @p N in each dimension.
+     */
+    template <int dim, typename CoefficientType>
+    void
+    resize(Table<dim, CoefficientType> &coeff, const unsigned int N)
+    {
+      TableIndices<dim> size;
+      for (unsigned int d = 0; d < dim; d++)
+        size[d] = N;
+      coeff.reinit(size);
+    }
+
+
+
+    /**
+     * we will need to take the maximum
+     * absolute value of fourier coefficients which correspond to $k$-vector
+     * $|{\bf k}|= const$. To filter the coefficients Table we
+     * will use the FESeries::process_coefficients() which requires a predicate
+     * to be specified. The predicate should operate on TableIndices and return
+     * a pair of <code>bool</code> and <code>unsigned int</code>. The latter
+     * is the value of the map from TableIndicies to unsigned int.  It is
+     * used to define subsets of coefficients from which we search for the one
+     * with highest absolute value, i.e. $l^\infty$-norm. The <code>bool</code>
+     * parameter defines which indices should be used in processing. In the
+     * current case we are interested in coefficients which correspond to
+     * $0 < i*i+j*j < N*N$ and $0 < i*i+j*j+k*k < N*N$ in 2D and 3D,
+     * respectively.
+     */
+    template <int dim>
+    std::pair<bool, unsigned int>
+    predicate(const TableIndices<dim> &ind, const unsigned int max_degree)
```

Comment:
A "predicate" is a function that returns a boolean. So this function is a predicate (of sorts), but the name doesn't say what criterion it applies in determining whether a table and max degree satisfies the predicate. Can you come up with a better name?

Reply:
I chose `predicate_k_in_r` for now.
Since this is part of an anonymous namespace, I wouldn't turn too much attention towards it.